Migrate to GitHub Actions

jacobtomlinson opened the pull request:
Closes #6787

This PR migrates CI to GitHub Actions.

To do:
- [x] Run tests on all os/python combinations
- [x] Run nightly tests on upstream repos
- [x] Import checks
- [x] Mindeps checks
- [x] HDFS checks

Changes:
- Coverage is only run on linux py38
- All python versions are now tested on macos

jacobtomlinson commented:
I *think* this is ready now. 

The mindeps and upstream tests are failing on my fork, but the failures are consistent with the existing failures in Travis. This should probably be investigated but is orthogonal to this migration.

jrbourbeau commented on .github/workflows/ci.yml:
```diff
@@ -1,34 +1,46 @@
-      run: |
-        cd docs
-        make html
+      - name: Coverage
+        shell: bash -l {0}
+        if: ${{ matrix.os == 'ubuntu-latest' &&  matrix.python-version == '3.8' }}
+        env:
+          COVERAGE: "true"
+          PARALLEL: "false"
+        run: source continuous_integration/scripts/coverage.sh
```
Currently we're getting a `No data to report.` error when this runs. We might consider switching over to using [Coverall's GitHub Action](https://github.com/marketplace/actions/coveralls-github-action). Additionally, while coverage is nice to have, I'd be okay commenting this out for now and addressing it in a follow-up PR if it means we can merge this sooner

jrbourbeau commented on .github/workflows/ci-upstream.yml:
```diff
@@ -0,0 +1,35 @@
+on:
+  schedule:
+    - cron: "0 1 * * *"
+  push:
+  pull_request:
```
Currently these CI builds will run on pushes to forks. We should consider only running on pushes / pull requests to the main branch.

jrbourbeau commented on .github/workflows/ci-upstream.yml:
```diff
@@ -0,0 +1,35 @@
+jobs:
+  build:
+    runs-on: "ubuntu-latest"
+    if: "contains(github.event.head_commit.message, 'test-upstream') || github.event_name != 'pull_request'"
```
I'm not 100% sure this is working as the upstream build was skipped for e5e2cd9872c3833897c8ec87bf905d2299bcd456 which includes "test-upstream" in the commit message

jacobtomlinson replied:
It only works for the first commit in the PR. I spent some time trying to figure out if you could trigger this on the latest commit on the branch but it doesn't seem the `github.event` object has this info.

jrbourbeau commented on continuous_integration/travis/install.sh:
```diff
@@ -1,54 +0,0 @@
-# We don't have a conda-forge package for cityhash
-# We don't include it in the conda environment.yaml, since that may
-# make things harder for contributors that don't have a C++ compiler
-python -m pip install --no-deps cityhash
```
Looks like we are no longer installing `cityhash` in the new `install.sh` script

jrbourbeau commented on .github/workflows/ci-additional.yml:
```diff
@@ -0,0 +1,79 @@
+      - name: Install
+        shell: bash -l {0}
+        env:
+          UPSTREAM_DEV: 1
```
Previously we weren't testing with `UPSTREAM_DEV` for the mindeps tests. As things can break with development versions of packages, we might consider not setting `UPSTREAM_DEV=1` here

jrbourbeau reviewed (commented):
Thanks for all your work on this @jacobtomlinson! I've left a few small-ish comments, but overall the changes here look good

jacobtomlinson commented:
Thanks @jrbourbeau I've addressed the review feedback.

I've commented out the coverage for now. I would actually propose that we switch to codecov in future. I've had good experience in other projects and it handles combining coverage very well. So you run coverage on everywhere and codecov merges the reports, this would mean things like the HDFS tests would be included in coverage on master.

But as you say it would be good to address this later.

gforsyth commented on .github/workflows/ci-additional.yml:
```diff
@@ -0,0 +1,77 @@
+        uses: actions/checkout@v2
+
+      - name: Setup Conda Environment
+        uses: goanpeca/setup-miniconda@v1.7.0
```
```suggestion
        uses: conda-incubator/setup-miniconda@v1.7.0
```

`goanpeca` migrated their action to the `conda-incubator` org

gforsyth commented on .github/workflows/ci-additional.yml:
```diff
@@ -0,0 +1,77 @@
+        uses: actions/checkout@v2
+
+      - name: Setup Conda
+        uses: goanpeca/setup-miniconda@v1.7.0
```
```suggestion
        uses: conda-incubator/setup-miniconda@v1.7.0
```

gforsyth commented on .github/workflows/ci-upstream.yml:
```diff
@@ -0,0 +1,35 @@
+        uses: actions/checkout@v2
+
+      - name: Setup Conda Environment
+        uses: goanpeca/setup-miniconda@v1.7.0
```
```suggestion
        uses: conda-incubator/setup-miniconda@v1.7.0
```

gforsyth commented on .github/workflows/ci.yml:
```diff
@@ -1,34 +1,43 @@
-      with:
-        python-version: 3.7
+      - name: Setup Conda Environment
+        uses: goanpeca/setup-miniconda@v1.7.0
```
```suggestion
        uses: conda-incubator/setup-miniconda@v1.7.0
```

gforsyth reviewed (commented):
Small changes for the migration from `goanpeca` -> `conda-incubator` org

jrbourbeau reviewed (approved):
Thanks for all your work on this @jacobtomlinson!

jacobtomlinson commented:
Thanks for the tweaks @gforsyth!
